pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -990,9 +1000,253 @@ func (t *Test) plan() (plan *TestPlan, retErr error) {
+		if v.Series() == msvSeries {
+			if v.LessThan(msv) {
+				return errors.Newf("minimum supported version is not <= %s", msv, v)
+			}
+		}
+		if v.Series() == mbvSeries {
+			if mbv != nil && v.LessThan(mbv) {
+				return errors.Newf("minimum bootstrap version is not <= %s", mbv, v)
+			}
+		}
+	}
+	return nil
+}
+
+// Check the following properties,
+// -- upgrade path satisfies the `<` version ordering.
+// -- skip upgrade when the upgrade path allows skip upgrade(s) and skipVersions is true.
+// -- no skip upgrade when skipVersions is false.
+// -- minimum supported version <= "from" version of a skip upgrade.
+// -- if skip upgrades are present, then the skip to final version is present.
+func (t *Test) assertSkipUpgrade(plan *TestPlan, skipVersions bool) error {
+	upgrades := plan.allUpgrades()
+	lastUpgrade := upgrades[len(upgrades)-1]
+	numSkips := 0
+	numSkipsToFinal := 0
+	hasEligibleSkips := false
+	{
+		if n, err := release.MajorReleasesBetween(&t.options.minimumSupportedVersion.Version, &lastUpgrade.to.Version); err != nil {
+			return errors.NewAssertionErrorWithWrappedErrf(err, "unexpected error in MajorReleasesBetween")
+		} else {
+			// If msv is 2 major releases from final version, and final version supports skip upgrades, then we expect a skip.
+			hasEligibleSkips = supportsSkipUpgradeTo(lastUpgrade.to) && n >= 2
+		}
+	}
+	var prev *upgradePlan
+
+	for i := 0; i < len(upgrades); i++ {
+		if !upgrades[i].from.Version.LessThan(upgrades[i].to.Version) {
+			return errors.Newf("invalid upgrade: %s -> %s", upgrades[i].from, upgrades[i].to)
+		}
+		if prev != nil {
+			if !prev.to.Version.Equals(upgrades[i].from.Version) {
+				return errors.Newf("invalid upgrade path: %s -> %s -> %s -> %s", prev.from, prev.to, upgrades[i].from, upgrades[i].to)
+			}
+		}
+		if n, err := release.MajorReleasesBetween(&upgrades[i].from.Version, &upgrades[i].to.Version); err != nil {
+			return errors.NewAssertionErrorWithWrappedErrf(err, "unexpected error in MajorReleasesBetween")
+		} else if n >= 2 {
+			numSkips++
+			if i == len(upgrades)-1 {
+				numSkipsToFinal++
+			}
+			if !skipVersions {
+				return errors.Newf("upgrade path: %s -> %s has a skip while skipProbability=%.2f", upgrades[i].from, upgrades[i].to, t.options.skipVersionProbability)
+			}
+			if upgrades[i].from.LessThan(t.options.minimumSupportedVersion) {
+				return errors.Newf("minimum supported version %s is not <= upgrade from %s to %s", t.options.minimumSupportedVersion, upgrades[i].from, upgrades[i].to)
+			}
+		}
+		prev = upgrades[i]
+	}
+	if skipVersions && hasEligibleSkips && numSkips == 0 {
+		return errors.Newf("expected at least one skip upgrade, got none")
+	}
+	if numSkips > 0 && supportsSkipUpgradeTo(upgrades[len(upgrades)-1].to) && numSkipsToFinal == 0 {
+		return errors.Newf("expected at least one skip upgrade to the final version, got none")
+	}
+	return nil
+}
+
+// assertExpectedUserHooks checks that the user-specified hooks occur in the plan.
```

Comment:
Could you refresh my memory? Was there a regression related to this, or just a nice to have check?

Reply:
Latter, I was just getting defensive now that we have explicit asserts during plan generation. I don't believe we've actually seen a case in the wild.